Chunk reuse code failing on multi-dimensional array (PyInf#11487)

Here's some code that seems to fail with the changes in https://github.com/deshaw/versioned-hdf5/commit/7662655467ec987e2efb28abde968b6a8cb7ad13:
```
import h5py
import numpy
import shutil
import tempfile
from   versioned_hdf5           import VersionedHDF5File

if __name__ == '__main__':
    # create temp dir path to be cleaned up on exit
    dir = tempfile.mkdtemp()
    path = f'{dir}/data.h5'
    try:
        rs = numpy.random.RandomState(0)
        dims = 3
        with h5py.File(path, 'w') as f:
            vf = VersionedHDF5File(f)
            with vf.stage_version('v0') as sv:
                for i in range(dims):
                    sv.create_dataset(f'axis{i}', dtype=numpy.dtype('int64'), shape=(0,),
                                      chunks=(10000,), maxshape=(None,))
                sv.create_dataset('value', dtype=numpy.dtype('int64'),
                                  shape=tuple([0 for _ in range(dims)]),
                                  chunks=tuple([20 for _ in range(dims)]),
                                  maxshape=tuple([None for _ in range(dims)]),
                                  fillvalue=0)
                sv.create_dataset('mask', dtype=numpy.dtype('int8'),
                                  shape=tuple([0 for _ in range(dims)]),
                                  chunks=tuple([20 for _ in range(dims)]),
                                  maxshape=tuple([None for _ in range(dims)]),
                                  fillvalue=2)
        for i in range(1, 101):
            with h5py.File(path, 'r+') as f:
                vf = VersionedHDF5File(f)
                with vf.stage_version(f'v{i}') as sv:
                    new_axes = tuple(numpy.unique(rs.randint(30, size=5))
                                     for _ in range(dims))
                    new_value = numpy.full(tuple(len(ax) for ax in new_axes), i)
                    new_mask = numpy.full(tuple(len(ax) for ax in new_axes), 0)

                    # figure out how existing axes map to new axes
                    new_indices = []
                    existing_indices = []
                    new_shape = []
                    for i in range(dims):
                        axis_ds = sv[f'axis{i}']
                        all_axis, indices = numpy.unique(numpy.concatenate([axis_ds[:], new_axes[i]]),
                                                         return_inverse=True)
                        existing_indices.append(tuple(indices[:len(axis_ds)]))
                        new_indices.append(tuple(indices[len(axis_ds):]))
                        axis_ds.resize((len(all_axis),))
                        axis_ds[:] = all_axis
                        new_shape.append(len(all_axis))

                    new_indices = tuple(new_indices)
                    existing_indices = tuple(existing_indices)
                    new_shape = tuple(new_shape)

                    # merge value
                    value_ds = sv['value']
                    all_data = numpy.full(new_shape, value_ds.fillvalue)
                    existing_data = value_ds[:]
                    all_data[numpy.ix_(*existing_indices)] = existing_data
                    all_data[numpy.ix_(*new_indices)] = new_value
                    value_ds.resize(new_shape)
                    value_ds[:] = all_data

                    # merge mask
                    mask_ds = sv['mask']
                    all_mask = numpy.full(new_shape, mask_ds.fillvalue)
                    existing_mask = mask_ds[:]
                    all_mask[numpy.ix_(*existing_indices)] = existing_mask
                    all_mask[numpy.ix_(*new_indices)] = new_mask
                    mask_ds.resize(new_shape)
                    mask_ds[:] = all_mask
    finally:
        shutil.rmtree(dir)
```
which fails for me with
```
/usr/local/bin/python3.11 /codemill/bessen/opensource/versioned-hdf5/append_insert.py 
Traceback (most recent call last):
  File "/codemill/bessen/opensource/versioned-hdf5/versioned_hdf5/backend.py", line 306, in _verify_new_chunk_reuse
    assert_array_equal(reused_chunk, to_be_written)
  File "/usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/testing/_private/utils.py", line 985, in assert_array_equal
    assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
  File "/opt/python/python-3.11/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/testing/_private/utils.py", line 778, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

(shapes (20, 20, 20), (20, 20, 6) mismatch)
 x: array([[[0, 2, 2, ..., 2, 2, 2],
        [0, 2, 2, ..., 2, 2, 2],
        [2, 2, 2, ..., 2, 2, 2],...
 y: array([[[0, 2, 2, 2, 2, 0],
        [0, 2, 2, 2, 2, 2],
        [2, 2, 2, 2, 2, 2],...

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/codemill/bessen/opensource/versioned-hdf5/append_insert.py", line 33, in <module>
    with vf.stage_version(f'v{i}') as sv:
  File "/opt/python/python-3.11/lib64/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/codemill/bessen/opensource/versioned-hdf5/versioned_hdf5/api.py", line 307, in stage_version
    commit_version(
  File "/codemill/bessen/opensource/versioned-hdf5/versioned_hdf5/versions.py", line 162, in commit_version
    slices = write_dataset(
             ^^^^^^^^^^^^^^
  File "/codemill/bessen/opensource/versioned-hdf5/versioned_hdf5/backend.py", line 194, in write_dataset
    _verify_new_chunk_reuse(
  File "/codemill/bessen/opensource/versioned-hdf5/versioned_hdf5/backend.py", line 308, in _verify_new_chunk_reuse
    raise ValueError(
ValueError: Hash b'\x80|\xebF\x95\xb4-\xaa\xc3\xdf\x95\x98\xd7\x94w\x8c\xc8|\xc2\xe3\x1e\xeet\xf9\xe0\xd1i\x18\xc6\x94\x7f\x95' of existing data chunk [[[0 2 2 ... 2 2 2]
  [0 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [0 2 2 ... 2 2 2]]

 [[2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]]

 [[2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]]

 ...

 [[2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]]

 [[0 0 2 ... 2 2 2]
  [0 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 0 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]]

 [[2 0 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]
  ...
  [2 2 2 ... 2 2 2]
  [2 0 2 ... 2 2 2]
  [2 2 2 ... 2 2 2]]] matches the hash of new data chunk [[[0 2 2 2 2 0]
  [0 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 2 2 0]
  [2 2 2 2 2 0]
  [0 2 2 2 2 2]]

 [[2 2 2 2 2 2]
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]]

 [[2 2 2 2 2 2]
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 0 0 2]
  [2 2 2 0 0 2]
  [2 2 2 2 2 2]]

 ...

 [[2 2 2 2 2 2]
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 0 0 2]
  [2 2 2 0 0 2]
  [2 2 2 2 2 2]]

 [[0 0 2 2 2 0]
  [0 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 2 2 2]
  [2 0 2 2 2 0]
  [2 2 2 2 2 2]]

 [[2 0 2 2 2 0]
  [2 2 2 2 2 2]
  [2 2 2 2 2 2]
  ...
  [2 2 2 2 2 2]
  [2 0 2 2 2 0]
  [2 2 2 2 2 2]]], but data does not.

Process finished with exit code 1
```

_Originally posted by @ArvidJB in https://github.com/deshaw/versioned-hdf5/issues/327#issuecomment-2150902625_

--- a/versioned_hdf5/backend.py
+++ b/versioned_hdf5/backend.py
@@ -2,7 +2,7 @@ import logging
 from typing import Dict, Optional
 
 import numpy as np
-from h5py import VirtualLayout, VirtualSource, h5s
+from h5py import VirtualLayout, VirtualSource, h5s, Dataset
 from h5py._hl.filters import guess_chunk
 from ndindex import ChunkSize, Slice, Tuple, ndindex
 from numpy.testing import assert_array_equal
@@ -189,7 +189,7 @@ def write_dataset(
                     slices[data_slice] = hashed_slice
 
                     _verify_new_chunk_reuse(
-                        raw_data=ds,
+                        raw_dataset=ds,
                         new_data=data,
                         data_hash=data_hash,
                         hashed_slice=hashed_slice,
@@ -227,11 +227,11 @@ def write_dataset(
 
 
 def _verify_new_chunk_reuse(
-    raw_data: np.ndarray,
+    raw_dataset: Dataset,
     new_data: np.ndarray,
     data_hash: bytes,
-    hashed_slice: Tuple,
-    chunk_being_written: Tuple,
+    hashed_slice: Slice,
+    chunk_being_written: np.ndarray,
     slices_to_write: Optional[Dict[Slice, Tuple]] = None,
     data_to_write: Optional[Dict[Slice, np.ndarray]] = None,
 ):
@@ -247,13 +247,13 @@ def _verify_new_chunk_reuse(
 
     Parameters
     ----------
-    raw_data : np.ndarray
-        Raw data that already exists in the file
+    raw_dataset : Dataset
+        Raw Dataset that already exists in the file
     new_data : np.ndarray
         New data that we are writing
     data_hash : bytes
         Hash of the new data chunk
-    hashed_slice : Tuple
+    hashed_slice : Slice
         Slice that is stored in the hash table for the given data_hash. This is a slice
         into the raw_data for the dataset; however if the data has not yet been written
         it may not point to a valid region in raw_data (but in that case it _would_
@@ -282,11 +282,11 @@ def _verify_new_chunk_reuse(
         # The hash table contains a slice that was written in a previous
         # write operation; grab that chunk from the existing raw data
         reused_slice = Tuple(
-            hashed_slice, *[slice(0, size) for size in new_data.shape[1:]]
+            hashed_slice, *[slice(0, size) for size in chunk_being_written.shape[1:]]
         )
-        reused_chunk = raw_data[reused_slice.raw]
+        reused_chunk = raw_dataset[reused_slice.raw]
 
-    # In some cases type coersion can happen during the write process even if the dtypes
+    # In some cases type coercion can happen during the write process even if the dtypes
     # are the same - for example, if the raw_data.dtype == dtype('O'), but the elements
     # are bytes, and chunk_being_written.dtype == dtype('O'), but the elements are
     # utf-8 strings. For this case, when the raw_data is changed, e.g.
@@ -371,7 +371,7 @@ def write_dataset_chunks(f, name, data_dict):
                     slices[chunk] = hashed_slice
 
                     _verify_new_chunk_reuse(
-                        raw_data=raw_data,
+                        raw_dataset=raw_data,
                         new_data=data_s,
                         data_hash=data_hash,
                         hashed_slice=hashed_slice,

--- a/versioned_hdf5/tests/test_api.py
+++ b/versioned_hdf5/tests/test_api.py
@@ -2742,3 +2742,80 @@ def test_dataset_getitem_can_read_direct(tmp_path, can_read_direct, expected_cal
                 )
 
         assert mock_crd.call_count == expected_calls
+
+
+def test_insert_in_middle_multi_dim(tmp_path):
+    """
+    Test we correctly handle inserting into a multi-dimensional Dataset
+    and shift the existing entries back.
+    """
+    rs = np.random.RandomState(0)
+    dims = 3
+    path = tmp_path / "tmp.h5"
+
+    with h5py.File(path, 'w') as f:
+        vf = VersionedHDF5File(f)
+        with vf.stage_version('v0') as sv:
+            for i in range(dims):
+                sv.create_dataset(f'axis{i}', dtype=np.dtype('int64'), shape=(0,),
+                                  chunks=(10000,), maxshape=(None,))
+            sv.create_dataset('value', dtype=np.dtype('int64'),
+                              shape=tuple([0 for _ in range(dims)]),
+                              chunks=tuple([20 for _ in range(dims)]),
+                              maxshape=tuple([None for _ in range(dims)]),
+                              fillvalue=0)
+            sv.create_dataset('mask', dtype=np.dtype('int8'),
+                              shape=tuple([0 for _ in range(dims)]),
+                              chunks=tuple([20 for _ in range(dims)]),
+                              maxshape=tuple([None for _ in range(dims)]),
+                              fillvalue=2)
+    for i in range(1, 101):
+        with h5py.File(path, 'r+') as f:
+            vf = VersionedHDF5File(f)
+            with vf.stage_version(f'v{i}') as sv:
+                new_axes = tuple(np.unique(rs.randint(30, size=5))
+                                 for _ in range(dims))
+                new_value = np.full(tuple(len(ax) for ax in new_axes), i)
+                new_mask = np.full(tuple(len(ax) for ax in new_axes), 0)
+
+                # figure out how existing axes map to new axes
+                new_indices = []
+                existing_indices = []
+                new_shape = []
+                for i in range(dims):
+                    axis_ds = sv[f'axis{i}']
+                    all_axis, indices = np.unique(np.concatenate([axis_ds[:], new_axes[i]]),
+                                                  return_inverse=True)
+                    existing_indices.append(tuple(indices[:len(axis_ds)]))
+                    new_indices.append(tuple(indices[len(axis_ds):]))
+                    axis_ds.resize((len(all_axis),))
+                    axis_ds[:] = all_axis
+                    new_shape.append(len(all_axis))
+
+                new_indices = tuple(new_indices)
+                existing_indices = tuple(existing_indices)
+                new_shape = tuple(new_shape)
+
+                # merge value
+                value_ds = sv['value']
+                all_data = np.full(new_shape, value_ds.fillvalue)
+                existing_data = value_ds[:]
+                all_data[np.ix_(*existing_indices)] = existing_data
+                all_data[np.ix_(*new_indices)] = new_value
+                value_ds.resize(new_shape)
+                value_ds[:] = all_data
+
+                # merge mask
+                mask_ds = sv['mask']
+                all_mask = np.full(new_shape, mask_ds.fillvalue)
+                existing_mask = mask_ds[:]
+                all_mask[np.ix_(*existing_indices)] = existing_mask
+                all_mask[np.ix_(*new_indices)] = new_mask
+                mask_ds.resize(new_shape)
+                mask_ds[:] = all_mask
+
+        with h5py.File(path, 'r') as f:
+            vf = VersionedHDF5File(f)
+            cv = vf[vf.current_version]
+            assert_equal(cv['value'], all_data)
+            assert_equal(cv['mask'], all_mask)

--- a/versioned_hdf5/tests/test_hashtable.py
+++ b/versioned_hdf5/tests/test_hashtable.py
@@ -273,3 +273,24 @@ def test_verify_chunk_reuse_strings(tmp_path):
         vf = VersionedHDF5File(f)
         with vf.stage_version("r1") as group:
             group["values"] = np.concatenate((data, data))
+
+
+def test_verify_chunk_reuse_multidim_1(tmp_path):
+    """Check that we correctly handle chunk reuse verification for multi-dimensional Datasets."""
+    filename = tmp_path / "testdata.h5"
+    with h5py.File(filename, mode="w") as f:
+        vf = VersionedHDF5File(f)
+        with vf.stage_version("r0") as group:
+            group.create_dataset(
+                "values",
+                data=np.array([[i + (j % 3) for i in range(8)] for j in range(7)]),
+                maxshape=(None, None),
+                chunks=(3, 3),
+            )
+
+    with h5py.File(filename, mode="r+") as f:
+        vf = VersionedHDF5File(f)
+        with vf.stage_version("r1") as group:
+            values_ds = group["values"]
+            values_ds.resize((8, 8))
+            values_ds[:] = np.array([[i + (j % 3) for i in range(8)] for j in range(8)])

